KHR_materials_transmission (#1698)
KhronosGroup/glTF: extensions/2.0/Khronos/KHR_materials_transmission/README.md

<figure>
  <img src="./figures/TransmissionTexture.png"/>
<figcaption><em>Controlling transmission amount with transmissionTexture.</em></figcaption>
</figure>

## Implementing Transmission ##

Modeling transmission is relatively straightforward. From the [glTF BRDF](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#appendix-b-brdf-implementation), we have:

*f* = *f*<sub>*diffuse*</sub> + *f*<sub>*specular*</sub>
*f*<sub>*diffuse*</sub> = (1 - *F*) * *diffuse*
*f*<sub>*specular*</sub> = *F* * *G* * *D* / (4 * dot(*N*, *L*) * dot(*N*, *V*))
, where *F* is the Surface Reflection Ratio.

*f* = *f*<sub>*diffuse*</sub> + *f*<sub>*specular*</sub>

Optical transparency does not require any changes whatsoever to the specular term so we only want to modify the diffuse term to account for transmitted light as well. We can do this by blending *f*<sub>*diffuse*</sub> with the transmitted light, **L**<sub>transmitted</sub>, based on the `transmission` values.

*f* = mix(**f**<sub>*diffuse*</sub>, **L**<sub>transmitted</sub> * `baseColor`, **T**) + *f*<sub>*specular*</sub>

> Where **L**<sub>transmitted</sub> is the colour of light coming through the material, towards the eye. It is tinted (multiplied) by the `baseColor`, and already has refraction taken into account (as discussed below). *T* is the transmission value defined by this extension.

Reviewer: How is **L**<sub>transmitted</sub> computed? I guess it includes (1-F)?

Author: I don't have a description of **L**<sub>transmitted</sub> in here yet but, yes, it would have to include 1-F. The IOR for that calculation would come from the base material (or an extension that overrides the base F0 of 0.04).

Author: I've cleaned up the math a bit here.

## Transparent Metals

The metallic parameter of a glTF material effectively scales the `baseColor` of the material toward black while, at the same time scaling the F0 (reflectivity) value towards 1.0. This effectively makes the material opaque for metallic values of 1.0 because transmitted light is attenuated out by absorption. This leaves only the specular component.

## Modeling Absorption

Absorption is usually defined as an amount of light at each frequency that is absorbed over a given distance through a medium (usually described by Beer’s Law). However, since this extension deals exclusively with very thin surfaces, we can treat absorption as a constant. In fact, rather than absorbed light, we can talk about its inverse: transmitted light. The `baseColor` of the material serves this purpose as it already defines how the light that penetrates the surface is colored by the material. In this model, the transmitted light will be modulated by this color as it passes through.

<figure>
  <img src="./figures/ConstantTransmission.png"/>
<figcaption><em>The baseColor of the material is used to tint the light being transmitted.</em></figcaption>
</figure>

### Modeling Refraction

> Since the surface is considered to be infinitely thin, we end up with a non-physically correct model. One the one hand, the material has no volume so light shouldn't be refracted when passing through. However, it should be expected that microfacets in a rough surface would refract the transmitted light in such a way as to blur the transmitted image. `How do we define this, mathematically??? By its nature, it's non-physically-based.`

Reviewer: Do you consider the effect of the index of refraction on thin refraction? E.g., as the index of refraction approaches 1, does the blur effect disappear?

Author: Yes, I think IOR should affect the amount of blur but I wasn't sure how to define anything mathematically yet. How much does an infinitely-thin surface refract light? Well, none...

There would be a (1-F) affect from light hitting the micro-facets, I guess, but no refraction.

Unless I'm not thinking about this correctly.

Author: Am I right in assuming that we do want a refractive blurring effect, even with an infinitely thin surface?
What do offline renderers do in this case? That should be our target reference.

For real-time, we can provide a recommended implementation but I think we need a higher-quality target so that real-time renderers are free to set their sights higher.

Reviewer: > How much does an infinitely-thin surface refract light?

If I am not mistaken, it depends on the orientation of the microfacets on the two sides of the object enclosing the infinitesimal small volume. If the facets on both sides are aligned such that the light ray entering the thin "volume" exists through a surface that has exactly the same orientation, there is no refraction. On the other hand, if the orientation is different, refraction will occur (if IOR != 1). This can be approximated with a microfacet lobe that is centered around the reflection direction flipped to the lower hemisphere wrt. the macro-surface normal. Here is an example how it is implemented in MDL for offline rendering: [libbsdf, line 514](https://github.com/NVIDIA/MDL-SDK/blob/2019.1.4/src/mdl/jit/libbsdf/libbsdf.cpp#L514). Another good source regarding offline rendering is the [Thin Surface BSDF from Imageworks (slide 34)](https://blog.selfshadow.com/publications/s2017-shading-course/imageworks/s2017_pbs_imageworks_slides_v2.pdf). They also show comparisons between actual modelled thin surfaces and their thin BTDF, and derive a roughness mapping that includes the IOR.

Author: Thank you. I've modified the spec to reflect this.

<figcaption><em>White sphere with alpha coverage of 25% (left) vs the same sphere with 100% optical transparency (right). Note that alpha coverage of 25% is shown because 0% would be completely invisible. Also note that alpha-as-coverage reflects light in proportion to its opacity. i.e. the more transparent it is, the weaker the reflections will appear.</em></figcaption>
</figure>

The `KHR_materials_transmission` extension provides a way to define glTF 2.0 materials that are transparent to light in a physically plausible way. That is, it enables the creation of transparent materials that absorb, reflect and transmit light depending on the incident angle and the wavelength of light. Common uses cases for thin-surface transmissive materials include plastics and glass.

This extension aims to address the simplest and most common use cases for optical transparency: infinitely-thin materials with no refraction, scattering, or dispersion. Dealing exclusively with “thin” materials (i.e. materials where only the surface is considered and not the volume) allows many simplifications to be made when calculating things like refraction and absorption.

## Extending Materials

> A transparent material is defined by adding the `KHR_materials_transmission` extension to any glTF material. When present, the extension indicates that a material should be rendered as a transparent surface and be blended as defined in the following spec. Note that, as rendering transparent objects presents many difficult-to-solve issues with primitive ordering, this extension does not dictate rendering algorithms.

Reviewer 2: I think that we should make a painters (depth sorting) algorithm normative - otherwise we risk objects 'popping' in and out.

Reviewer 3: Mike is correct that specific rendering techniques are out of scope for normative glTF spec..  This particular suggestion would prohibit many real-time implementations, which is not what we want.

Separately, 3D Commerce may require something along these lines for certification, but beware that you may be excluding more clients than you bargained for here.

Reviewer 2: I think it could pose a performance problem for a particular type of real-time implementations - mostly webGL / script based ones.
The cost would be performance - however the gain would be predictable results.
Depending on implementation (language and platform) you would get varying degrees of performance penalty, for instance a C based engine could do it on the CPU without too much overhead.
A Vulkan based engine could do it on the GPU.

I would not want us to limit this due to current viewer implementation bottlenecks - it could be a good motivator to push development forward.

Reviewer 4: The rendering should not be normative.
Depending on the rendering method (forward vs. deferred plus it's variations), different and better algorithms do exist.
E.g. pixel sorting (aka order independent transparency) should not be excluded.
From my point of view, we should point to non-normative approaches.

Reviewer 3: @rsahlin Also consider that many glTF implementations exist within rendering libraries that are used by 3rd-party applications.  The library cannot dictate how the application structures its render passes or loops.  That's part of what makes this extension particularly challenging, as it imposes some nontrivial best practices there.  These should be as clearly spelled out as we can manage, but not marked normative.  Little more than the glTF file structure and its meaningful interpretation is normative to glTF.

Reviewer 2: I think it is a simplification calling this the 'rendering ' - to me it boils down to what is visible through the primitives using transmission.

If it's the sorting algorithm you object to then I would suggest a wording similar to:
'The background and other objects, including primitives using transmission, should be visible through the material (that is using transmission)'

Reviewer 5: As discussed on today's call, Sketchfab's refraction material has a similar baseline as what is proposed here as an implementation note: we only require the opaque scene to be sampled by the transmission material.  From their [documentation ](https://help.sketchfab.com/hc/en-us/articles/202602073-Transparency):
- A refractive object can not refract itself.
- A refractive object can not refract other refractive objects.
- There can be sorting issues when mixing refractive and transparent objects.
- There can be artifacts when the camera moves.
- Objects outside of the field of view are not refracted.

And for convenience, here are some [sample models](https://sketchfab.com/Sketchfab/collections/refraction-examples) that demonstrate their refraction material.

Sketchfab has established this as an acceptable baseline for their community and I would hope that our spec would be structured in such a way that their implementation is conformant.  It seems like a reasonable baseline.

Reviewer 6: Seems to be a similar situation for Blender's realtime renderer (Eevee). It cannot show refractive objects behind other refractive objects, but can display transparent and opaque objects behind refractive objects. https://blenderartists.org/t/realistic-glass-in-eevee/1149937/29

Note that alpha-as-coverage can still be used along with transmission as shown in the example below.
<figure>
  <img src="./figures/TransmissionWithMask.png"/>
<figcaption><em>Alpha coverage and optical transparency can be used at the same time so that some areas of a surface are transparent while others disappear entirely.</em></figcaption>
</figure>

## Transparent Metals
Metals effectively absorb all refracted light (light that isn't reflected), preventing transmission.
The metallic parameter of a glTF material effectively scales the `baseColor` of the material toward black while, at the same time scaling the F0 (reflectivity) value towards 1.0. This makes the material opaque for metallic values of 1.0 because transmitted light is attenuated out by absorption. Therefore, for a material with `metallicFactor=1.0`, the value of `transmissionFactor` doesn't matter.

> ## Implementing Transmission ##

Reviewer 2: Please make it clear that this is an implementation note:

> **Implementation Note:**

Author: This section is normative. The non-normative "Implementation Notes" are below this section.

We will now add an additional term for the transmitted light:

*f* = *f*<sub>*diffuse*</sub> + *f*<sub>*specular*</sub> + *f*<sub>*transmission*</sub>

*f*<sub>*transmission*</sub> = (1 - *F*) * *T* * *D<sub>T</sub>* * *baseColor*

where *T* is the transmission percentage defined by this extension's `transmission` and `transmissionTexture` properties and *D<sub>T</sub>* is the distribution function for the transmitted light. The distribution function is the same Trowbridge-Reitz model used by specular reflection except sampled along the view vector rather than the reflection. The *baseColor* factor causes the transmitted light to be tinted by the surface.

Light that penetrates a surface and is transmitted will not be diffusely reflected so we also need to modify the diffuse calculation to account for this.
*f*<sub>*diffuse*</sub> = (1 - *F*) * (1 - *T*) * *diffuse*

Optical transparency does not require any changes whatsoever to the specular term. So this can all be rearranged so that the transmitted light amounts to a modifaction of the diffuse lobe.
*f* = (1 - *F*) * (*T* * *D<sub>T</sub>* + (1 - *T*) * *diffuse*) + *f*<sub>*specular*</sub>

> ## Implementation Notes

Reviewer 2: I'm leaning towards that we should require a painter algorithm (depth sorting) of triangles - in my opinion this should be normative.  
Making depth sorting of triangles normative will make a huge step towards consistent result across viewers (bar blurring and possibly other effects)
Without it whole objects may pop in and out of existence......

However, I'm not sure how to approach blurring since I think this will be the most 'heavy' effect to implement for multiple layers/objects.

Author: Depth sorting of triangles is not going to be reasonable on all platforms due to the number of passes required (or having the ability to use linked-lists of fragments which isn't possible in WebGL). Besides, I think rendering techniques are out of scope of glTF. Yes, there will be differences between renderers but as long as they do the blending in the way described, I think that's enough.
Blurring of the transmission is pretty cheap if they just sample the mips of the background texture. The background render texture will probably be required anyway to do the appropriate blending.

Reviewer 6: I agree that sorting triangles is prohibitively expensive. Maybe a dedicated model viewer (note: model viewers are a subset of glTF clients) can do that when the camera and object are stationary, but it's not a requirement we should have in the extension spec.

But on a similar point — the core specification says this about alpha blending:

> Whether depth value is written or whether to sort is up to the implementation.

In retrospect, I think we could have made a normative requirement that depth _not_ be written for alpha blend materials. Does that same decision apply to transmission? Or is this not an issue, because transmissive objects are treated as opaque (or rendered in an entirely separate pass)?

Reviewer 2: > Depth sorting of triangles is not going to be reasonable on all platforms due to the number of passes required (or having the ability to use linked-lists of fragments which isn't possible in WebGL).

I agree that if all layers have to support blurring an engine could require multiple passes (render to texture, minify then sample) which could become a performance bottleneck.
However - 'just' sorting the triangles of transmission objects could be done in a single pass if effects needing to sample the objects behind are left out.
This would leave us with viewers that would at least display transparent objects in the correct order and visibility.

Yes, the blurring itself is quite cheap - it's having to render out to texture, forcing multiple passes, that will be mostly heavy.

I also do not like us to be so tied to WebGL - just because something can not be done in WebGL should, in my opinion, not mean that we leave it out.

Reviewer 4: Hope "blurring" is just in short we do it physically - as far as possible - correct.

Reviewer 2: > Hope "blurring" is just in short we do it physically - as far as possible - correct.

Sure, that's what I mean anyway. :-)

The point I'm trying to make is that we should require a minimum of visibility through triangles using transmission.
Meaning that not only the background should show through - any other primitives (that may also use transmission) should also be visible.
However, what effects to apply to each of these primitives is optional - for instance the roughness that will result in blurring (since will require multiple passes on todays hardware)

Reviewer 7: > I also do not like us to be so tied to WebGL - just because something can not be done in WebGL should, in my opinion, not mean that we leave it out.

I agree with this in principle but WebGL will have a huge influence since we want glTF to be viewable on the web.

Reviewer 6: I don't think this is a WebGL problem. Requiring that implementations sort triangles, in realtime, is extremely burdensome on any platform. Yes, there are cases where it may be possible (i.e. few enough triangles), but we really cannot expect this at the data format level.

Is there a reason these requirements cannot be in the 3D Commerce Certification process, instead?

# KHR\_materials\_transmission

## Contributors

* Mike Bond, Adobe, [@miibond](https://github.com/MiiBond)

## Status

Work in Progress

> ## DependenciesI

Reviewer 3: `DependenciesI` -> `Dependencies`

Author: Thanks. Finally addressed this (only 3 months late).

|   |Type|Description|Required|
|---|----|-----------|--------|
|**transmissionFactor** | `number` | The base percentage of light that is transmitted through the surface.| No, default:`1.0`|
|**transmissionTexture** | [`textureInfo`] | A greyscale texture that defines the transmission percentage of the surface. This will be multiplied by transmissionFactor. | No |

### transmissionFactor 

The amount of light that is transmitted by the surface rather than diffusely re-emitted. This is a percentage of all the light that penetrates a surface (i.e. isn’t specularly reflected) rather than a percentage of the total light that hits a surface. A value of 1.0 means that 100% of the light that penetrates the surface is transmitted through. This is the default.

<figure>
  <img src="./figures/ConstantTransmission.png"/>
<figcaption><em>Defaulting to 1.0 allows making a material transparent by simple defining the extension.</em></figcaption>
</figure>

### transmissionTexture 

> A greyscale texture that defines the amount of light that is transmitted by the surface rather than absorbed and re-emitted. A value of 1.0 means that 100% of the light that penetrates the surface (i.e. isn’t specularly reflected) is transmitted through. The value is linear and is multiplied by the transmissionFactor to determine the total transmission value. Note that use of transmission values does not disallow usage of alpha coverage (via `baseColor.a`).

Reviewer 8: Could it be a single channel selected for this? If it's red, we could pair it with the metallic roughness texture.

Author: The red channel of the metallic-roughness texture can already be used for ambient occlusion.
However, I had thought that we could share this texture with the KHR_materials_volume (#1726) extension which could use one of the other channels for thickness information (if we go the route of allowing the thickness param).

Author: Dispersion (the prism effect) is specifically not part of this extension to keep things simple. It could certainly be included in another extension though.

Reviewer 6: I agree with @jackcaron that we should pick a single channel, for two reasons:

1. By requiring that three channels be filled with the same value, we are losing any packing flexibility that might (intentionally or not) have been available. The fact that the <kbd>r</kbd> channel in the ORM texture could be used for either AO *or* transmission is a good thing.
2. If in the future we have a general-purpose swizzle extension, it will be awkward that it's required to provide a 3-channel output for a 1-channel value here.

Either <kbd>r</kbd> or <kbd>a</kbd> seem like reasonable choices, to me.

Reviewer 9: I would propose this extension use the r channel of a texture separate from the ORM, as I would not want to have to omit occlusion in order to take advantage of per-pixel transmission. Pre-calculated ambient occlusion has a strong visual benefit.

Perhaps we could use a channel of another texture, and reserve the green and blue channels for other PBR Next features? Transmission (r), clearcoatTexture (g), and clearcoatRoughnessTexture (b) for example.

Reviewer 6: We don't need to choose a specific texture to pair transmission with — as long as the channels don't conflict, users can mix and match. Any of (transmission; occlusion/rough/metal), (occlusion; transmission/rough/metal), (transmission; occlusion; ____/rough/metal) would be allowed.

Author: We've been discussing using the 'G' channel for thickness over in the [KHR_materials_volume](https://github.com/KhronosGroup/glTF/pull/1726) thread.

Reviewer 2: I agree that the texture should be one channel (prism not being part of this extension).
As for which channel to choose I would propose either R or A.
Picking A would make it easy to have occlusion/roughness/metallic/transmission (without swizzle)
Picking R could have benefit if occlusion is less likely to be used with this extension.

Either way is fine to me, I will solve by internally swizzling the textures if needed.

# KHR\_materials\_transmission

## Contributors

* Mike Bond, Adobe, [@miibond](https://github.com/MiiBond)

## Status

Work in Progress

## Dependencies

Written against the glTF 2.0 spec.

## Overview

> Many optically transparent materials are impossible to represent in a physically plausible manner with the core glTF 2.0 PBR material. This is because the core specification only includes the concept of "alpha as coverage" (exposed via the alpha channel of `baseColorTexture`). Many common materials like glass and plastic require significantly different modeling.

Reviewer 6: > Also, I believe this has already been discussed, but should this extension be renamed to thin_transmission?

I think I prefer `KHR_materials_transmission`. While this extension _alone_ models only thin surfaces, it is also the foundation for future volume transparency extensions, which will use information in this extension. In other words, this extension does not indicate that a surface is thin: that is the default state. Other extensions are required to make a surface "thick."

## Extending Materials

A transparent material is defined by adding the `KHR_materials_transmission` extension to any glTF material. When present, the extension indicates that a material should be rendered as a transparent surface and be blended as defined in the following spec. Alpha coverage and `doubleSided` properties still apply to transparent materials. Note that, as rendering transparent objects presents many difficult-to-solve issues with primitive ordering, this extension does not dictate rendering algorithms.

```json
materials: [
  {
    "extensions": {
       "KHR_materials_transmission": {
         "transmissionFactor": 0.8,
         "transmissionTexture": 0
       }
    }
  }
]
```

## Properties 

Only two properties are introduced with this extension and combine to describe a single value; the percentage of light that is transmitted through the surface of the material. These properties work together with the existing properties of the material to define the way light is modified as it passes through the substance. 

|   |Type|Description|Required|
|---|----|-----------|--------|

> |**transmissionFactor** | `number` | The base percentage of light that is transmitted through the surface.| No, default:`1.0`|

Reviewer 3: Somehow this review came out of sync with the latest version, but the default has been updated to the correct value.

|**transmissionTexture** | [`textureInfo`] | A greyscale texture that defines the transmission percentage of the surface. This will be multiplied by transmissionFactor. | No |

### transmissionFactor 

> The amount of light that is transmitted by the surface rather than diffusely re-emitted. This is a percentage of all the light that penetrates a surface (i.e. isn’t specularly reflected) rather than a percentage of the total light that hits a surface. A value of 1.0 means that 100% of the light that penetrates the surface is transmitted through. This is the default.

Reviewer 3: This part didn't get updated, and still shows the old default value.  So this line still needs a correction.

> The amount of light that is transmitted by the surface rather than diffusely re-emitted. This is a percentage of all the light that penetrates a surface (i.e. isn’t specularly reflected) rather than a percentage of the total light that hits a surface. A value of 1.0 means that 100% of the light that penetrates the surface is transmitted through. This is the default.

Author: Thanks. Fixed.

<figure>
  <img src="./figures/ConstantTransmission.png"/>
<figcaption><em>Defaulting to 1.0 allows making a material transparent by simple defining the extension.</em></figcaption>
</figure>

### transmissionTexture 

> A greyscale texture that defines the amount of light that is transmitted by the surface rather than diffusely re-emitted. A value of 1.0 means that 100% of the light that penetrates the surface (i.e. isn’t specularly reflected) is transmitted through. The value is linear and is multiplied by the transmissionFactor to determine the total transmission value. Note that use of transmission values does not disallow usage of alpha coverage (via `baseColor.a`).

Reviewer 2: If alpha coverage is not needed could we have a solution where the transmissionTexture  scalar is taken from baseColor.a?

This would optimize use of samplers and bandwidth for many usecases.

Reviewer 3: I'm a little torn on this.  If there were a good fallback path to be had here, where non-transmission-aware renderers could do the best fallback by finding the transmission values mixed into the alpha coverage channel, maybe that would be a reason to try this?

But for non-fallback behavior, this seems like a problem.  For example Blender keeps these values separate for a reason, and it's possible to use both on the same material.

Reviewer 6: Yes. Use of both alpha blend and transmission seems sketchy to me, but there may be reasons to do so. Use of both alpha *mask* and transmission seems totally reasonable. I think it would be unsafe for this extension to try to alter its behavior based on the alphaMode (and as @emackey points out, Blender avoids that too) so I think the extension is probably correct as written.

Reviewer 6: You _could_ make an argument that the `transmissionTexture`'s single channel should be `a` instead of `r`? But I'm having trouble convincing myself whether that's a good idea... using `baseColorTexture` and `transmissionTexture` with the same texture would cause problems in some engines — it is difficult to really ensure that a 0% alpha channel doesn't have leak-through side effects even when the material is marked as opaque. Especially if that texture is being reused elsewhere _with_ the alpha channel...

Author: I think these two things are very different concepts and we shouldn't do anything to confuse them for the user. Besides, using baseColor.a for transmission would require taking the inverse of the value which further confuses things.

> A greyscale texture that defines the amount of light that is transmitted by the surface rather than diffusely re-emitted. A value of 1.0 means that 100% of the light that penetrates the surface (i.e. isn’t specularly reflected) is transmitted through. The value is linear and is multiplied by the transmissionFactor to determine the total transmission value. Note that use of transmission values does not disallow usage of alpha coverage (via `baseColor.a`).

Reviewer 2: Multiplication of texture and scale value is standard for glTF - however, do we see the need to keep this.
Ie, will the texture be used with transmissionFactor != 1.0?

Reviewer 4: If one want to use transmission without a texture, one can use the `transmissionFactor` to set the value per material. In this case, someone does not need to store the `transmissionTexture`, which does save a lot of resources.
Using the `transmissionTexture` without a factor makes sense from your side, so no need to deep dive into this case.
To have both in combination does allow to have e.g. variations. Same texture, the scalar is doing the variation. Also, if both defined, the calculation somehow needs to make sense.

In general, this is glTF style and I do not see any reason to change this. It always have been a `logical or` and there is no need to switch to an `either or`. @lexaknyazev

The effort to implement this in a shader is minimal and the effort grows linearly and not exponentially.
Finally, maybe the scalars/vectors are animated in the future.

So, at this point of time, this discussion is out of scope.

Reviewer 3: I agree, this discussion feels out of scope here.  @rsahlin, you can detect `*Factor == 1.0` for any glTF factor at the app level, and use a shader compiler directive to get rid of the extra multiply, if you believe it's problematic.  But in practice I don't think it's something we need to shift strategy on for future extensions.

Reviewer 2: > The effort to implement this in a shader is minimal and the effort grows linearly and not exponentially.
> Finally, maybe the scalars/vectors are animated in the future.

@UX3D-nopper My question is who uses the scale factor != 1.0 together with a texture.
Not how it is declared in the specification or how to implement in shader.

So, the answer is:
>In general, this is glTF style and I do not see any reason to change this.

That's fine!

Reviewer 2: @emackey Yes, I do know how to handle shader variants using one super-shader :-)

>But in practice I don't think it's something we need to shift strategy on for future extensions.

Sure, that's fine.

<figure>
  <img src="./figures/TransmissionTexture.png"/>
<figcaption><em>Controlling transmission amount with transmissionTexture.</em></figcaption>
</figure>

## Implementing Transmission ##

From the core [glTF BRDF](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#appendix-b-brdf-implementation), we have:

*f* = *f*<sub>*diffuse*</sub> + *f*<sub>*specular*</sub>

> *f*<sub>*diffuse*</sub> = (1 - *F*) * *diffuse*

Reviewer 5: There is some odd formatting here in the rendered result (newlines are missing).

Reviewer 3: Newlines are strange in markdown.  You actually need two trailing spaces at the end of a line to force a newline.

Alternately, you could place a blank line between the two formulas.

Author: Oh, weird. You're right. I was previewing it in VSCode with a plugin and it displayed it fine.
I'll add some newlines.

*f*<sub>*specular*</sub> = *F* * *G* * *D* / (4 * dot(*N*, *L*) * dot(*N*, *V*))
, where *F* is the Surface Reflection Ratio.

We will now add an additional term for the transmitted light:

*f* = *f*<sub>*diffuse*</sub> + *f*<sub>*specular*</sub> + *f*<sub>*transmission*</sub>

*f*<sub>*transmission*</sub> = (1 - *F*) * *T* * *D<sub>T</sub>* * *baseColor*

> where *T* is the transmission percentage defined by this extension's `transmission` and `transmissionTexture` properties and *D<sub>T</sub>* is the distribution function for the transmitted light. The distribution function is the same Trowbridge-Reitz model used by specular reflection except sampled along the view vector rather than the reflection. The *baseColor* factor causes the transmitted light to be tinted by the surface.

Reviewer 5: I personally would benefit from clarity here with regards to `Dt`.  This states that `Dt` is the Towbridge-Reitz distribution function, but sampled along the view direction rather than the reflected direction.  The microfacet distribution function, D, is calculated using dot(n, h).  Is it correct that this should instead use dot(n, -v) in this calculation for when we deal with transmission? Is it that simple?

Author: I think the `dot(n, h)` term shouldn't come into play at all since the transmitted light isn't being reflected, but I could be wrong. The incoming light direction should be sampled along the view direction because you're looking through the material and are seeing completely unrefracted light. The transmitted light is then scaled by the Fresnel to account for the portion of light reflected off the inside surface of the material (and, therefore, not reaching your eye).
Not sure at all if this is correct so it would be good to have others (like @proog128) look at this.

Reviewer 10: To my understanding, the dot product is relevant as we still evaluate the microfacet distribution. There's a brief section on the topic in the DSPBR spec https://bit.ly/2NpmJVn (eq. 5 and fig. 4). We just use a modified half-vector, which uses the flipped light direction.

Author: Yes, you're right. You'd flip it across the plane defined by the surface normal and then it would be centered around the same direction as the view.

Reviewer 5: In this sentence you mention Trowbridge-Reitz by name, but we only suggest that model in Appendix B as non-normative.  Should it be mentioned here?  I would consider rewriting this as:
```suggestion
where *T* is the transmission percentage defined by this extension's `transmission` and `transmissionTexture` properties and *D<sub>T</sub>* and *G<sub>T</sub>* are the microfacet distribution and shadowing functions for the transmitted light. These *D<sub>T</sub>* and *G<sub>T</sub>* functions are the same models used by specular reflection, except sampled along the view vector rather than the reflection. The *baseColor* factor causes the transmitted light to be tinted by the surface.
```

Light that penetrates a surface and is transmitted will not be diffusely reflected so we also need to modify the diffuse calculation to account for this.
*f*<sub>*diffuse*</sub> = (1 - *F*) * (1 - *T*) * *diffuse*

Optical transparency does not require any changes whatsoever to the specular term. So this can all be rearranged so that the transmitted light amounts to a modifaction of the diffuse lobe.
*f* = (1 - *F*) * (*T* * *D<sub>T</sub>* + (1 - *T*) * *diffuse*) + *f*<sub>*specular*</sub>

> 

Reviewer 5: Should there be guidance on how transmission works with IBL?  A hint here would be useful: When dealing with light contributions from an IBL source, your IBL environment map is sampled in the same manner as with the surface reflection, but for transmission we use the view vector rather than the reflected view vector.  

I assume that is how it works.

## Realtime Implementation Notes

Implementations of transmission are expected to vary, especially in real-time engines, depending on the capabilities of the runtime, but should endeavor to achieve plausibility. 

### Refractive Blurring
Some rasterizer implementations may opt to sample prefiltered IBL or dynamic light probes to simulate refractive blurring due to surface roughness. Others may choose to sample from a blurred version of the rendered scene behind a surface. Raytracers are expected to use more physically correct refraction with a BSDF supporting multi-scattering like the modified [Smith](https://eheitzresearch.wordpress.com/240-2/) model.

<figure>
  <img src="./figures/Roughness.png"/>
<figcaption><em>Refraction due to surface roughness.</em></figcaption>
</figure>

> **TODO** - *This is fine for raytracers or for rasterizers that just sample the existing IBL for refracted light. However, what about rasterizers that want to cheaply render transmission in screen-space (I think this is pretty common)? Typically, the refractive blurring is done by sampling mips or manually-blurred versions of the background scene. This can be eye-balled by an implementation but is there a slightly more mathematically-rigorous version that we could recommend?*

Reviewer 5: I do think it would be really helpful here to include some best practices on how rasterizers can produce plausible background blurring through rough surfaces.  Are there good references that you can provide here?  What are other rasterizers doing for this?

Author: I actually removed all of this. I wonder if it belongs in the spec. or perhaps in a different doc?

### Blending

> The glTF `blendMode` is used for alpha-as-coverage, NOT for physically-based transparency (i.e. this extension). If alpha-as-coverage is not being used, the blend mode of the material should be set to "OPAQUE" even though it is transparent. Again, it's helpful to think of alpha-as-coverage as whether the physical surface is there or not. `transmission` applies to the surface material that exists.

Reviewer 5: I did notice that in the sample model, all materials are flagged as "BLENDED".  While on the topic of the sample model, it requires the KHR_texture_transform extension; that dependency should be removed.

Author: Good catch. Only the ones with opacity should have had that. This is fixed now.

Reviewer 3: Since the sample model has appeared in https://github.com/KhronosGroup/glTF-Sample-Models/pull/265, can the sample files associated with this PR be removed?  I'd like for there not to be two sources of sample models (although I realize some other extensions already violate this).

Author: Will do.

## Limitations of Alpha as Coverage
Alpha-as-coverage can basically be thought of as describing whether the surface exists or not. When alpha is 0, nothing is rendered while alpha=1 means that the material is solid. It's most useful for materials such as gauze or burlap that only partially cover the area in which the material is rendered. These materials contain lots of small holes where the light can pass through so, strictly speaking, the light isn't actually entering the material itself (i.e. the surface isn't there). For other materials, like clear glass, a photon of light might actually enter the surface and be transmitted through. In this case, the light can be reflected, refracted, absorbed or scattered. Simple alpha-as-coverage cannot represent any of these effects and so we need very different logic to render them correctly.

The most important of these features is surely reflection. Even a completely transparent, infinitely thin, glass material is still visible due to its reflection of light. However, a material using purely alpha as coverage is not visible at all when alpha = 0 (because the surface effectively doesn't exist).

<figure>

> <img src="./figures/OpacityComparison.png"/>

Reviewer 5: I just noticed that there appears to be some refraction in this reference image.  Does this need to be updated?

Author: Updated. Thanks for noticing that.

Only two properties are introduced with this extension and combine to describe a single value; the percentage of light that is transmitted through the surface of the material. These properties work together with the existing properties of the material to define the way light is modified as it passes through the substance. 

|   |Type|Description|Required|
|---|----|-----------|--------|
|**transmissionFactor** | `number` | The base percentage of light that is transmitted through the surface.| No, Default: `0.0 ` |
|**transmissionTexture** | [`textureInfo`](/specification/2.0/README.md#reference-textureInfo) | A texture that defines the transmission percentage of the surface, stored in the `R` channel. This will be multiplied by `transmissionFactor`. | No |

### transmissionFactor 
The amount of light that is transmitted by the surface rather than diffusely re-emitted. This is a percentage of all the light that penetrates a surface (i.e. isn’t specularly reflected) rather than a percentage of the total light that hits a surface. A value of 1.0 means that 100% of the light that penetrates the surface is transmitted through.

<figure>
  <img src="./figures/ConstantTransmission.png"/>

> <figcaption><em>baseColor or yellow with transmissionFactor = 1.0.</em></figcaption>

Reviewer 11: I think "baseColor of" is meant instead of "baseColor or". 

I also noticed that the example model specifies a magenta baseColorFactor for the model that appears yellowish, which is presumably due to the phrasing in this spec that says:

    rather than absorbed light, we can talk about its inverse: transmitted light

So, I think this needs to be clarified.  Should we subtract baseColor from (1,1,1) ?  Should the caption really say "magenta"?

Author: Good catch. Thank you.
It's fixed now.

<figure>
  <img src="./figures/ConstantTransmission.png"/>
<figcaption><em>Yellow baseColor with transmissionFactor = 1.0.</em></figcaption>
</figure>

### transmissionTexture 
The `R` channel of this texture defines the amount of light that is transmitted by the surface rather than diffusely re-emitted. A value of 1.0 in the red channel means that 100% of the light that penetrates the surface (i.e. isn’t specularly reflected) is transmitted through. The value is linear and is multiplied by the transmissionFactor to determine the total transmission value.

<figure>
  <img src="./figures/TransmissionTexture.png"/>
<figcaption><em>Controlling transmission amount with transmissionTexture.</em></figcaption>
</figure>

## Tinting
You may notice that the transparent materials shown above are tinted. i.e. they aren't actually transmitting 100% of the non-reflected light. This is because `KHR_materials_transmission` specifies that the `baseColor` of the material be used to model the absorption of light by the surface. This allows materials like stained glass to be easily represented with this extension.

> Absorption is usually defined as an amount of light at each frequency that is absorbed over a given distance through a medium (usually described by Beer’s Law). However, since this extension deals exclusively with infinitely thin surfaces, we can treat absorption as a constant. In fact, rather than absorbed light, we can talk about its inverse: transmitted light. The `baseColor` of the material serves this purpose as it already defines how the light that penetrates the surface is colored by the material. In this model, the transmitted light will be modulated by this color as it passes through.

Reviewer 11: Normally, the `baseColorTexture` uses the sRGB transfer function and must be converted to linear space before it is used for any computations.  When the baseColorTexture is used tinting, should we assume that this sRGB transformation still applies?  This should perhaps be clarified.

Author: Yes, the transform should definitely still apply. I can clarify in the specs. Thanks.

Reviewer 3: Let's not confuse `baseColor` with `baseColorTexture` or `baseColorFactor` here.  Only `baseColorTexture` is supplied as sRGB, the `baseColorFactor` is supplied already as linear.  The final `baseColor` is a product of the two linear values, after conversion.

> ## Tinting
> You may notice that the transparent materials shown above are tinted. i.e. they aren't actually transmitting 100% of the non-reflected light. This is because `KHR_materials_transmission` specifies that the `baseColor` of the material be used to model the absorption of light by the surface. This allows materials like stained glass to be easily represented with this extension.
> Absorption is usually defined as an amount of light at each frequency that is absorbed over a given distance through a medium (usually described by Beer’s Law). However, since this extension deals exclusively with infinitely thin surfaces, we can treat absorption as a constant. In fact, rather than absorbed light, we can talk about its inverse: transmitted light. The `baseColor` of the material serves this purpose as it already defines how the light that penetrates the surface is colored by the material. In this model, the transmitted light will be modulated by this color as it passes through. Note that because `baseColor` is defined as sRGB, it must be converted to linear before being used in this calculation.

Reviewer 3: Actually can I propose a bit of a rewrite?  I'm concerned we've buried the lede here.

```suggestion
## Tinting
The `baseColor` of the material, as defined in the [Metallic-Roughness Material](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#metallic-roughness-material)
section of the glTF specification, defines the amount of light at each frequency
that is transmitted (not absorbed) by the material. Absorption is usually defined
as an amount of light at each frequency that is absorbed over a given distance
through a medium (usually described by Beer’s Law). However, since this
extension deals exclusively with infinitely thin surfaces, we can treat
absorption as a constant. Therefore, its inverse, `baseColor` transmission,
can also be constant.  This provides the color tinting effect in the materials
shown above.
```

Reviewer 6: I like Ed's the rephrasing, but two nits — in the last two sentences could you replace the words "its" and "This" with the terms they refer to? I had a bit of trouble tracking the pronouns at the end.

Author: I rephrased this a bit. Maybe it's clearer now?

### transmissionFactor 
The amount of light that is transmitted by the surface rather than diffusely re-emitted. This is a percentage of all the light that penetrates a surface (i.e. isn’t specularly reflected) rather than a percentage of the total light that hits a surface. A value of 1.0 means that 100% of the light that penetrates the surface is transmitted through.

<figure>
  <img src="./figures/ConstantTransmission.png"/>
<figcaption><em>Yellow baseColor with transmissionFactor = 1.0.</em></figcaption>
</figure>

### transmissionTexture 
The `R` channel of this texture defines the amount of light that is transmitted by the surface rather than diffusely re-emitted. A value of 1.0 in the red channel means that 100% of the light that penetrates the surface (i.e. isn’t specularly reflected) is transmitted through. The value is linear and is multiplied by the transmissionFactor to determine the total transmission value.

<figure>
  <img src="./figures/TransmissionTexture.png"/>
<figcaption><em>Controlling transmission amount with transmissionTexture.</em></figcaption>
</figure>

## Tinting

> The `baseColor` of the material, as defined in the [Metallic-Roughness Material](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#metallic-roughness-material) section of the glTF specification, defines the amount of light at each frequency that is transmitted (not absorbed) by the material. Absorption is usually defined as an amount of light at each frequency that is absorbed over a given distance through a medium (usually described by Beer’s Law). However, since this extension deals exclusively with infinitely thin surfaces, we can treat absorption as a constant. Therefore, its inverse, `baseColor` transmission, can also be constant.  This provides the color tinting effect in the materials shown above and is useful for real-world materials like stained glass and tinted plastics. Note that because `baseColor` is defined as sRGB, it must be converted to linear before being used in this calculation.

Reviewer 3: The note at the end, "baseColor is defined as sRGB", is not correct.

Don posted a handy diagram explaining baseColor here:  https://github.com/KhronosGroup/glTF/issues/1638#issuecomment-668918360

Author: Oh, good catch. I'll just remove that statement completely.

</figure>

## Transparent Metals
Metals effectively absorb all refracted light (light that isn't reflected), preventing transmission.
The metallic parameter of a glTF material effectively scales the `baseColor` of the material toward black while, at the same time scaling the F0 (reflectivity) value towards 1.0. This makes the material opaque for metallic values of 1.0 because transmitted light is attenuated out by absorption. Therefore, for a material with `metallicFactor=1.0`, the value of `transmissionFactor` doesn't matter.

## Transmission BTDF ##
From the core [glTF BRDF](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#appendix-b-brdf-implementation), we have:

*f* = *f*<sub>*diffuse*</sub> + *f*<sub>*specular*</sub>

*f*<sub>*diffuse*</sub> = (1 - *F*) * *diffuse*

*f*<sub>*specular*</sub> = *F* * *G* * *D* / (4 * dot(*N*, *L*) * dot(*N*, *V*)), where *F* is the Surface Reflection Ratio.

We will now add an additional term for the transmitted light:

*f* = *f*<sub>*diffuse*</sub> + *f*<sub>*specular*</sub> + *f*<sub>*transmission*</sub>

> *f*<sub>*transmission*</sub> = (1 - *F*) * *T* * *D<sub>T</sub>* * *baseColor*

Reviewer: Isn't this missing the G term and the normalization, like in f_specular?

f_transmission = (1-F) * T * baseColor * D_T * G_T / (4 * abs(dot(N, L)) * abs(dot(N, V)))

Author: Yes, quite right. Thank you.

> *f*<sub>*transmission*</sub> = (1 - *F*) * *T* * *baseColor* * *D<sub>T</sub>* * *D<sub>G</sub>* / *(4 * abs(dot(N, L))* * *abs(dot(N, V)))*

Reviewer 5: Minor, but I think you meant G<sub>T</sub>.  May want to also mention something about G<sub>T</sub> below.  There's a related conversation about this term on the [KHR_materials_volume PR](https://github.com/KhronosGroup/glTF/pull/1726#discussion_r465386060)

where *T* is the transmission percentage defined by this extension's `transmission` and `transmissionTexture` properties and *D<sub>T</sub>* is the distribution function for the transmitted light. The distribution function is the same Trowbridge-Reitz model used by specular reflection except sampled along the view vector rather than the reflection. The *baseColor* factor causes the transmitted light to be tinted by the surface.

Light that penetrates a surface and is transmitted will not be diffusely reflected so we also need to modify the diffuse calculation to account for this.
*f*<sub>*diffuse*</sub> = (1 - *F*) * (1 - *T*) * *diffuse*

Optical transparency does not require any changes whatsoever to the specular term. So this can all be rearranged so that the transmitted light amounts to a modifaction of the diffuse lobe.

> *f* = (1 - *F*) * (*T* * *D<sub>T</sub>* + (1 - *T*) * *diffuse*) + *f*<sub>*specular*</sub>

Reviewer 5: Typo: modifaction.

I think this section needs to get updated to include the additions of G<sub>T</sub> / (4.0 * NdotL * NdotV) from above.

> ## Implementation Notes

Reviewer 6: Consider adding a sub-header...

> *This section is non-normative.*

... similar to what we've done in the [appendices of the core spec](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#appendix-b-brdf-implementation).

</figure>

## Schema

- [glTF.KHR_materials_transmission.schema.json](schema/glTF.KHR_materials_transmission.schema.json)

## Reference

[Autodesk Standard Surface - Specular Transmission](https://autodesk.github.io/standard-surface/#closures/speculartransmission)  
[Blender Transparent BSDF](https://docs.blender.org/manual/en/latest/render/shader_nodes/shader/transparent.html#transparent-bsdf)  
[Enterprise PBR Shading Model - Transparency](https://dassaultsystemes-technology.github.io/EnterprisePBRShadingModel/spec-2020x.md.html)  
[Filament Material models - Transmission](https://google.github.io/filament/Materials.md.html#materialmodels/litmodel/transmission)  
[Unreal Engine 4 Material - Refraction](https://docs.unrealengine.com/en-US/Engine/Rendering/Materials/MaterialInputs/index.html#refraction)

> [Adobe Standard Material - Interior Properties](https://helpx.adobe.com/dimension/using/standard-materials.html#InteriorProperties)

Reviewer 5: Minor: The final MD puts the UE4 and Adobe links on the same line.

Reviewer 3: Fixed.
